### ISSUE TITLE
Changed Behaviour on Object-Construction on Phalcon 2.0.11+

### ISSUE THREAD

**Issue by jirgn** [phalcon/cphalcon#11805]
When upgrading from Phalcon 2.0.10 to 2.0.12 i experienced a problem on construction of a object derived from Phalcon/Http/Response in my microapplication.
In my code i found it handy to derive from Phalcon/Http/Response like 
    
    class MyResponse extends Phalcon/Http/Response 
    {
        function __construct($myOwn, $sinature)    {
            // no parent constructor called here

            $this->setContentType();
        }
        //some further implementation
    }

I see it is no clean code at all and therefore no issue for the most of you, but when instantiating an instance like

    $myResponseInstance = new MyResponse("some", "param");

I got an exception like "Trying to call method set on a non-object" triggered by the method setContentType. Also when removing this function call it seems that the object has never been created.
The simple fix for this was calling the **parent::__constuct()** method. 
I think this behaviour is not expectable and if present in future releases, this behaviour should be described in the documentation.

hope I help some one who also stumbles uppon this

greez jirgn

**Comment by stamster** [comment]
In general, 
you should use ```onConstruct()``` instead of ```__construct()```.

**Comment by Skydev0h** [comment]
Of course there will be errors if you override constructor and do not call parent constructor which indeed does some initialization for the class.
Response constructor does some initialization, to be more precise
```
...
let this->_headers = new Headers();
...
```
And `setContentType` in turn calls `setHeader` which does
```
...
let headers = this->getHeaders();
headers->set(name, value);
...
```
And guess what `getHeaders` returns? Right! `return this->_headers;`!

TL;DR: Never override native `__construct` without calling `parent::_construct` first in it! :D

**Comment by stamster** [comment]
@Skydev0h 
Well, this used to work prior to v2.0.11 w/o calling parent constructor.

But I do agree that constructor should not be over-driven. Instead use custom Setter.

**Comment by jirgn** [comment]
When writing this issue, i just wanted to point out, that you might have breaking changes in your 2.0.11+ releases. I also mentioned, that the code is no a good practice at all, but it would be a good thing, to add a hint in the change logs, or make a general hint in the documentation how to handle inheritance in the phalcon world.

**Comment by stamster** [comment]
Actually wherever you extended native Phalcon class w/o calling parent constructo it might break your code.
